src/main/scala/latis/util/hapi/HapiBinaryEncoder.scala

```diff
@@ -0,0 +1,40 @@
+package latis.util.hapi
+
+import scodec._
+import scodec.bits._
+
+import latis.data.Data
+import latis.data.Data._
+import latis.model._
+import latis.output.BinaryEncoder
+
+object HapiBinaryEncoder {
+  private def int2ByteVal(x:Int):ByteValue = ByteValue(x.toByte)
+  private def int2ShortVal(x:Int):ShortValue = ShortValue(x.toShort)
+  private def double2FloatVal(x:Double):FloatValue = FloatValue(x.toFloat)
+
+  def hapiCodec(s: Scalar): Codec[Data] = s.valueType match {
+    case BooleanValueType => codecs.fail(Err("HapiBinaryEncoder does not support Boolean."))
+    case ByteValueType    => codecs.int32L.xmap[ByteValue](int2ByteVal, _.value.toInt).upcast[Data]
+    case ShortValueType   => codecs.int32L.xmap[ShortValue](int2ShortVal, _.value.toInt).upcast[Data]
+    case IntValueType     => codecs.int32L.xmap[IntValue](IntValue, _.value).upcast[Data]
+    case LongValueType    => codecs.fail(Err("HapiBinaryEncoder does not support Long."))
+    case FloatValueType   => codecs.doubleL.xmap[FloatValue](double2FloatVal, _.value.toDouble).upcast[Data]
```

Comment:
Now that you say it, I think that's right that float.toDouble can potentially introduce minor errors, so maybe we just shouldn't allow floats?

For the others, then would the thought be that any "integer" encoded from any of the types would always have to be decoded as an `IntValue` rather than allowing decoding to `ShortValue` or `ByteValue`? And then have attempting to decode into either of those two result in an error?